### ISSUE TITLE
sudden cut at the end of the midi

### ISSUE THREAD

**Issue by kotc** [Mindwerks/wildmidi#116]
can the renderer not quit until all notes are done playing? (even if the midi score finished already)

**Comment by chrisisonwildcode** [comment]
What's needed in the player is polling of the drivers to see if they are finished parsing the data sent to them before quitting.

 Originally WildMIDI player worked flawlessly however "more experienced" developers sent in patches that distros were using to "do things the right way" and the player over time has evolved into unrecognisable evil.

I am tempted to take the player and return it to its former self where it worked as I intended and not how others felt it should. After all a game engine as the original test bed for libwildmidi. *ahh those were the days* ... Upside is that the live performance features could then be implemented and demonstrated ... Downside is people complaining about "that's not how you do it!".

/rant

**Comment by kotc** [comment]
im not talking about player part, the lib does that to me. my idea is to:
1/ let the score do it thing
2/ when the score finishes, let the notes play and wait for silence before WildMidi_GetOutput() returns 0 bytes?

**Comment by psi29a** [comment]
I'm pro player re-write. :)  For 0.4.3 then? hehehe

**Comment by kotc** [comment]
well, i need only lib fixed (and most likely player part will be fixed automatically because it uses same lib)

**Comment by chrisisonwildcode** [comment]
Which beach are you referring?

**Comment by psi29a** [comment]
beach? @kotc can you provide a midi and state which audio subsystem you are using? alsa, oss, openal?

**Comment by kotc** [comment]
im on vacations right now so im little limited in resources. as for player, its my own written from scratch using WildMidi_GetOutput() in a loop until it returns 0 then loads next tune and goes in a loop again. i'll grab the representative midi when i have a chance, but just grab any file that ends with an instrument that has longer decay (ie. some grand piano)

**Comment by psi29a** [comment]
So this issue has nothing to do with our player and it's audio driver, but with the output of libwildmidi. Thanks for the clarification.

**Comment by chrisisonwildcode** [comment]
@kotc which version are you using? And if you're using our development source, which branch ...

**Comment by kotc** [comment]
i think i got main 0.4.0 from the github (ref: refs/heads/master)

**Comment by psi29a** [comment]
0.4 is not yet released, and development is now in wildcode-updates branch. When that is finished, it will be merged into master and we'll have a release.

Please try again with the wildcode-updates branch.

**Comment by chrisisonwildcode** [comment]
please test -updates to see if this still occurs.

**Comment by kotc** [comment]
will do when i get a chance, but i bet its the way the WildMidi_GetOutput() works. how can i checkout/switch branch?

**Comment by chrisisonwildcode** [comment]
Kotc, I am curious about how you are using the lib ... you're not quite clear on you usage of it and your wording would suggest its not being used as intended.

basic sequence should be
WildMidi_Init
WildMidi_Open
WildMidi_GetOutput repeatedly until returns 0
WildMidi_Close
WildMidi_Shutdown

Between open and close you can call other WildMidi functions for changing options and to get info, etc ...

Does your player follow that sequence?

**Comment by chrisisonwildcode** [comment]
@kotc I think what you are referring to has been fixed in the wildmidi-updates branch. The fix is still being tested as it is automatic and dynamic depending on patches used and amount of time after the last note off event.

I am unclear about your usage of the lib however the bare bones lib usage (with the proper syntax of course) should be ...

WildMidi_Init
WildMidi_Open
WildMidi_GetOutput repeatedly until it returns 0
WildMidi_Close
WildMidi_Shutdown

You can call multiple Open/Close within a single Init/Shutdown

Note: wildmidi-updates is in active development and is may contain issues and/or may be unstable. Some features within wildmidi-updates may not make it into release.

**Comment by kotc** [comment]
yes, thats the sequence im using. and when switching song im calling close then open. i'll try the -updates branch it when i get a chance (which may take a few days). on unrelated note, i have a few midis that sound like they have their notes cut in the middle by playing next notes, how does wildmidi play consecutive notes on the same track (or whatever its called, in timidity there was polyphony setting for such case i think)

**Comment by chrisisonwildcode** [comment]
WildMidi can play all 127 notes at once per channel if someone really wanted to for some reason ...

As others have said, wildcode-updates has ALOT of changes compared to master and is the branch that most fixes and features are tested in. Any feedback about your experience with the wildcode-updates branch is welcomed and remember to update regularly.

At some point soon I expect there to be a 0.4.0 release direct from the wildcode-updates branch. I am not adding any more features to the branch at this time (although I snuck one in over night), mainly doing a lot of testing and refining where needed.

**Comment by chrisisonwildcode** [comment]
I should add that WildMidi will kill off a notes very quickly (but not instantly) under certain circumstances. These include

1: a note is already active (which can happen for a number of reasons) and the same note is requested on the same channel. WildMidi will kill off quickly (but not instantly) the active note and then replay it.
2: all notes off (or other events that silence notes) will kill off notes quickly 
3: some gus_pat envelopes will kill off notes quickly.

**Comment by psi29a** [comment]
@chrisisonwildcode just a heads up, the process is that after a branch is finished, we merge it into master. That branch is deleted. These long living branches do not work well with our development because it gets out of sink fast.  We should strive to do feature branches, meaning one feature per branch. Merge that in, then delete the branch. :)

**Comment by kotc** [comment]
hmm, this might be the point 1 (note already active on same channel). could there be an option for lib to dynamically create additional channels just for such case (up for some polyphony max setting) ?

**Comment by chrisisonwildcode** [comment]
First you need to try the newer source first. And if the problem still exists we can work out what's going on.

Unfortunately your suggestion of creating more channels is not how poly works in midi.

**Comment by kotc** [comment]
hmm, i've compiled updates-branch and i must say it wounds worse, much worse because notes are cut too fast, even faster than in main branch (i'm talking about notes inside song, havent' checked the end of midi case yet). im using "Beethoven_Sonata_14_Op_27_No_2_Moonlig
ht_mov_3.mid", size 40433, freepats. want me to upload this file somewhere?

**Comment by chrisisonwildcode** [comment]
Yes please upload the midi file and then provide a link. I'll grab freepats from their website

**Comment by psi29a** [comment]
@chrisisonwildcode To be fair, I noticed this too... at doom's gate guitar strokes sound cut-off in relation to old master. Maybe this is a preference thing, but I preferred the longer notes. It sounded better.

**Comment by kotc** [comment]
here it goes:
http://www.filedropper.com/oscptmp-beethovensonata14op27no2moonlightmov3
you can hear the difference in the first few seconds

**Comment by psi29a** [comment]
@chrisisonwildcode had a moment to look at this yet?

**Comment by chrisisonwildcode** [comment]
I am working on stuff that may affect this, will asked for it to be checked once it is done.

**Comment by psi29a** [comment]
@chrisisonwildcode you done here?

@kotc you consider this "good enough" in master?

**Comment by kotc** [comment]
so far so good for me, if i stumble upon midi that ends abruptly i'll reopen it

**Comment by psi29a** [comment]
sure :)
